src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/HttpUpgradeTransport.cs

```diff
@@ -0,0 +1,274 @@
+
+                using (var delayCts = new CancellationTokenSource())
+                {
+                    var resultTask = await Task.WhenAny(sending, Task.Delay(_options.CloseTimeout, delayCts.Token));
+
+                    if (resultTask != sending)
+                    {
+                        // We timed out so now we're in ungraceful shutdown mode
+                        Log.CloseTimedOut(_logger);
+
+                        // Abort the websocket if we're stuck in a pending send to the client
+                        _aborted = true;
+
+                        context.Abort();
+                    }
+                    else
+                    {
+                        delayCts.Cancel();
+                    }
+                }
+            }
+            else
+            {
+                Log.WaitingForClose(_logger);
+
+                // We're waiting on the websocket to close and there are 2 things it could be doing
+                // 1. Waiting for websocket data
+                // 2. Waiting on a flush to complete (backpressure being applied)
+
+                using (var delayCts = new CancellationTokenSource())
+                {
+                    var resultTask = await Task.WhenAny(receiving, Task.Delay(_options.CloseTimeout, delayCts.Token));
+
+                    if (resultTask != receiving)
+                    {
+                        // Abort the websocket if we're stuck in a pending receive from the client
+                        _aborted = true;
+
+                        context.Abort();
+
+                        // Cancel any pending flush so that we can quit
+                        _application.Output.CancelPendingFlush();
+                    }
+                    else
+                    {
+                        delayCts.Cancel();
+                    }
+                }
+            }
+        }
+
+        private async Task StartReceiving(Stream stream)
+        {
+            try
+            {
+                while (true)
+                {
+#if NETCOREAPP2_2
+                    // Do a 0 byte read so that idle connections don't allocate a buffer when waiting for a read
+                    await stream.ReadAsync(Memory<byte>.Empty, CancellationToken.None);
+#endif
+                    var memory = _application.Output.GetMemory();
+
+#if NETCOREAPP2_2
+                    var read = await stream.ReadAsync(memory, CancellationToken.None);
+#else
+                    var isArray = MemoryMarshal.TryGetArray<byte>(memory, out var arraySegment);
+                    Debug.Assert(isArray);
+
+                    // Exceptions are handled above where the send and receive tasks are being run.
+                    var read = await stream.ReadAsync(arraySegment.Array, arraySegment.Offset, arraySegment.Count, CancellationToken.None);
+#endif
+                    // Need to check again for NetCoreApp2.2 because a close can happen between a 0-byte read and the actual read
```

Comment:
Copy pasta.
So what happens if the 0-byte read returns from a close then we try the real read?
I think before when we didn't check it threw when you tried to read again, but I don't know the behavior of the stream